ckanext/xloader/loader.py

```diff
@@ -13,18 +13,29 @@
 from chardet.universaldetector import UniversalDetector
 from six.moves import zip
 from tabulator import config as tabulator_config, EncodingError, Stream, TabulatorException
+from tabulator.config import CSV_SAMPLE_LINES
 from unidecode import unidecode
 
 import ckan.plugins as p
 
 from .job_exceptions import FileCouldNotBeLoadedError, LoaderError
-from .parser import CSV_SAMPLE_LINES, TypeConverter, CanadaCSVParser
+from .parser import TypeConverter
 from .utils import datastore_resource_exists, headers_guess, type_guess
 
 from ckan.plugins.toolkit import config
 
 import ckanext.datastore.backend.postgres as datastore_db
 
+try:
+    from ckanext.canada.tabulate import (
+        CanadaStream as TabulatorStream,
+        CanadaCSVParser as TabulatorCSVParser
```

Comment:
This is a little magical. Why not have these classes in our xloader fork if this is the only way they're used?

Reply:
Similar comment here, if xloader had an interface to let us provide stream and csvparser that would be cleaner. Importing from a specific theme from a reusable code base just feels like the wrong direction.

Reply:
@wardi sorry, they were in our Xloader fork. But will be needing to use them in Goodtables and Validation

Reply:
Yeah, so if you give xloader, goodtables and validation interfaces that we could call during configuration of ckanext-canada to pass in the classes and config then we don't end up with our forks of these depending on our specific theme plugin in a fragile way.

This is another argument for building validation features into xloader to reduce the complexity, but maybe there are some users that need one without the other?

Reply:
@wardi okay so just something like `ITabulator` and have methods like `get_dialect`, `get_stream_class`, `get_parsers`, and `get_encoding`??

Would all three plugins (goodtables, Xloader, Validation) need to have this plugin implementation in their codebase? So just copy the `ITabulator` class to each one?

Reply:
It could be something as simple as a function you can call to change the default configuration/register the new classes required in goodtables, xloader and validation. Doesn't have to be the same on all three.

Reply:
@wardi https://github.com/open-data/ckanext-validation/pull/13/commits/68e6e49310b55fccc815463c42c8628b1438f0fe something like this??

Reply:
Or am I reaaaaly beneath monkey patching Tabulator classes in our Canada plugin hahaha

Reply:
that's an option, and one I've reached for in the past. Not a great option because monkey patching is super confusing for the person debugging a few years from now, but if it's extremely painful to implement an interface to allow validation to be configured I could understand. Just make sure to use lots of FIXME comments in ckanext-canada explaining why it was done that way :-/